src/mito2/src/compaction/memory_manager.rs

```diff
@@ -0,0 +1,597 @@
+    {
+        let raw = String::deserialize(deserializer)?;
+        let lower = raw.to_ascii_lowercase();
+
+        if lower == "skip" {
+            return Ok(OnExhaustedPolicy::Skip);
+        }
+        if lower == "wait" {
+            return Ok(OnExhaustedPolicy::default());
+        }
+        if lower.starts_with("wait(") && lower.ends_with(')') {
+            let inner = &raw[5..raw.len() - 1];
+            let timeout = parse_duration(inner).map_err(serde::de::Error::custom)?;
+            return Ok(OnExhaustedPolicy::Wait { timeout });
+        }
+
+        Err(serde::de::Error::custom(format!(
+            "invalid compaction memory policy: {}, expected wait, wait(<duration>) or skip",
+            raw
+        )))
+    }
+}
+
+/// Global memory manager for compaction tasks.
+#[derive(Clone)]
+pub struct CompactionMemoryManager {
+    quota: Option<MemoryQuota>,
+}
+
+/// Shared memory quota state across all compaction guards.
+#[derive(Clone)]
+struct MemoryQuota {
+    semaphore: Arc<Semaphore>,
+    // Maximum permits (aligned to PERMIT_GRANULARITY_BYTES).
+    limit_permits: u32,
+}
+
+impl CompactionMemoryManager {
+    /// Creates a new memory manager with the given limit in bytes.
+    /// `limit_bytes = 0` disables the limit.
+    pub fn new(limit_bytes: u64) -> Self {
+        if limit_bytes == 0 {
+            COMPACTION_MEMORY_LIMIT.set(0);
+            return Self { quota: None };
+        }
+
+        let limit_permits = bytes_to_permits(limit_bytes);
+        let limit_aligned_bytes = permits_to_bytes(limit_permits);
+        COMPACTION_MEMORY_LIMIT.set(limit_aligned_bytes as i64);
+
+        Self {
+            quota: Some(MemoryQuota {
+                semaphore: Arc::new(Semaphore::new(limit_permits as usize)),
+                limit_permits,
+            }),
+        }
+    }
+
+    /// Returns the configured limit in bytes (0 if unlimited).
+    pub fn limit_bytes(&self) -> u64 {
+        self.quota
+            .as_ref()
+            .map(|quota| permits_to_bytes(quota.limit_permits))
+            .unwrap_or(0)
+    }
+
+    /// Returns currently used bytes.
+    pub fn used_bytes(&self) -> u64 {
+        self.quota
+            .as_ref()
+            .map(|quota| permits_to_bytes(quota.used_permits()))
+            .unwrap_or(0)
+    }
+
+    /// Returns available bytes.
+    pub fn available_bytes(&self) -> u64 {
+        self.quota
+            .as_ref()
+            .map(|quota| permits_to_bytes(quota.available_permits_clamped()))
+            .unwrap_or(0)
+    }
+
+    /// Acquires memory, waiting if necessary until enough is available.
+    pub async fn acquire(&self, bytes: u64) -> CompactionMemoryGuard {
+        match &self.quota {
+            None => CompactionMemoryGuard::unlimited(),
+            Some(quota) => {
+                let permits = bytes_to_permits(bytes);
+                if permits == 0 {
+                    return CompactionMemoryGuard::unlimited();
+                }
```

Comment:
Do we need to ensure the permits < limit_permits? Maybe use min(limit_permits) to avoid allocating out of limit permits.

Reply:
Good catch! Permits should be rounded up to a minimum of 1; otherwise, the subsequent request_additional becomes unrestricted, which is a bug.

Reply:
fixed in [e40a15b](https://github.com/GreptimeTeam/greptimedb/pull/7305/commits/e40a15bdb731ae7838c34db216b66055d6a3e888)